### PR TITLE
[azcosmos] Add ClientOptions.DisableEndpointDiscovery to route all requests to the client endpoint

### PR DESCRIPTION
**Pull request by logangeorge01**
### Problem

We need to call a Cosmos account through an endpoint whose **domain name differs** from the one the service advertises. The account's readable/writable locations advertise a host (a different DNS suffix) that is not resolvable/routable from our client's network — all traffic must instead flow through the endpoint the client was constructed with (a reverse proxy on a different domain).

Today the SDK learns the advertised host during endpoint discovery and the client retry policy rewrites every request's host to it (`req.Raw().URL.Host = resolvedEndpoint.Host` in `cosmos_client_retry_policy.go`), so requests leave the reachable endpoint. There is currently no public option to keep traffic pinned to the endpoint the client was created with: `enableCrossRegionRetries` is hardcoded `true` in both `NewClient`/`NewClientWithKey`, and `ClientOptions` only exposes `PreferredRegions`.

> Note: it may be that readable/writable locations were only ever intended to carry multiple Cosmos *regions* (same DNS suffix), so advertising a different suffix per location may not be a supported topology — this is still to be confirmed on our side. Either way, an explicit client-side opt-out is useful when the account is fronted by a proxy on a different domain.

### Change

Add `ClientOptions.DisableEndpointDiscovery` (default `false`, so existing behavior is unchanged). When set to `true`:

- `locationCache.resolveServiceEndpoint` short-circuits to the client (default) endpoint for both reads and writes; endpoint discovery metadata still runs but its result is not used for routing.
- **Cross-region failover is suppressed** in the client retry policy while **same-region (same-endpoint) transport retries are preserved** — a transient DNS/connect blip at the pinned endpoint is still retried in-region. The `disableEndpointDiscovery` guard is applied only to the endpoint-changing paths: the cross-region section of `attemptRetryOnNetworkError`, `attemptRetryOnEndpointFailure`, `attemptRetryOnServiceUnavailable`, `attemptRetryOnRequestTimeout`, `attemptRetryOnSessionUnavailable`, and the cross-region branch of `attemptRetryOnServerError` (the in-region 5xx retry is preserved).
- `PreferredRegions` may still be set but has no effect on routing (matching the Python SDK).

This mirrors `EnableEndpointDiscovery=false` in the .NET/Java Cosmos SDKs and the equivalent Python behavior.

Wiring:
- `ClientOptions.DisableEndpointDiscovery bool` (new public field).
- Threaded through `NewClient` / `NewClientWithKey` -> `newGlobalEndpointManager` -> `newLocationCache` as the negative `disableEndpointDiscovery`. Kept negative so the zero value means discovery enabled, matching prior behavior for the existing `locationCache` literals that omit the field.

### Tests

- `TestResolveServiceEndpointWithEndpointDiscoveryDisabled` / `TestResolveServiceEndpointWithEndpointDiscoveryEnabled` — routing resolves to the client endpoint vs. the advertised regional endpoints.
- Constructor wiring: `TestNewClientWithKeyDisableEndpointDiscovery`, `TestNewClientDisableEndpointDiscovery`, `TestNewClientDisableEndpointDiscoveryAcceptsPreferredRegions`.
- `TestDisableEndpointDiscoveryKeepsSameRegionRetriesButSuppressesFailover` — a transient failure at the pinned endpoint recovers via same-region retries, and once that budget is exhausted the call fails fast without any failover request to another advertised region.
- `TestSessionNotAvailableDisableEndpointDiscoverySuppressesRetry` — a 404/1002 surfaces on the first attempt with no session-failover reroute.
- `go build`, `go vet ./...`, and the full `azcosmos` package test suite pass.

### Changelog

Added an entry under `1.6.0-beta.4 (Unreleased)`.

### CODE REVIEW

**Comment by Pilchie** [comment]
@simorenoh - can you take a look at the review?

@FabianMeiswinkel for the overall approach. Setting this option would prevent regional failover from working, right?

**Comment by simorenoh** [comment]
@logangeorge01 as @Pilchie mentioned, this approach means the SDK always targets the configured client endpoint, regardless of the account’s regions. The SDK would no longer perform regional routing or failover, so `PreferredRegions` would have no effect. Your reverse proxy would then be responsible for routing requests to another region during an outage.

**Comment by logangeorge01** [comment]
@Pilchie @simorenoh thanks for taking a look. To give a little more context, our product is Azure Local Disconnected Operations which runs cosmos airgapped on premises where there is only one 'region', so we don't have a need for PreferredRegions/routing. This change just aligns the go sdk with dotnet which already supports disabling endpoint discovery

File: sdk/data/azcosmos/cosmos_client.go [hunk 1]
```diff
@@ -82,12 +82,16 @@ func NewClientWithKey(endpoint string, cred KeyCredential, o *ClientOptions) (*C
 	}
 	preferredRegions := []string{}
 	enableCrossRegionRetries := true
+	enableEndpointDiscovery := true
 	if o != nil {
 		preferredRegions = o.PreferredRegions
+		if o.DisableEndpointDiscovery {
+			enableEndpointDiscovery = false
```
Comment (simorenoh) [inline sdk/data/azcosmos/cosmos_client.go]: When endpoint discovery is disabled, also disable cross-region retries and reject `PreferredRegions`. Otherwise, retries can behave as though another region was selected while endpoint resolution continues returning the same client endpoint, resulting in ineffective failover attempts.

**Comment by logangeorge01** [comment]
Thanks for the review! Pushed a commit (d3b1b89) addressing the feedback:

**Zero-value routing regression (valid, now fixed):** Switched the internal cache flag from a positive nableEndpointDiscovery to a negative disableEndpointDiscovery. This preserves zero-value behavior for the existing `locationCache{...}` literals in tests that omit the field (e.g. `makeGEMWithRegions` and the retry-policy fixtures), so discovery/regional routing stays on by default. This was a real failure — `TestReadServerError_CrossRegionRoutesToDifferentEndpoint` failed before this change and passes now.

**Constructor-level test coverage (valid):** Added `TestNewClientWithKeyDisableEndpointDiscovery` and `TestNewClientDisableEndpointDiscovery`. Both build a real client through the public constructors, load advertised regional endpoints into the cache, and assert that with `DisableEndpointDiscovery: true` reads and writes resolve to the client endpoint, while the default routes to an advertised regional endpoint. This guards against the option being dropped/inverted in either constructor path.

**Option doc wording (valid):** Clarified the `DisableEndpointDiscovery` doc comment to state that the account metadata request may still run and only its result is ignored for routing (the metadata call itself isn't disabled), and noted that regional failover is disabled so `PreferredRegions` has no effect.

**"Tests fail to compile" comments:** These were against the first commit (before the callers were updated) — all `newLocationCache` / `newGlobalEndpointManager` call sites were updated in a later commit, and the full package test suite (`go test ./...`) is green.

On the overall approach (@FabianMeiswinkel @simorenoh): agreed this pins routing to the client endpoint and disables failover — that's intended for our single-region disconnected/air-gapped scenario, and it mirrors the .NET/Java `EnableEndpointDiscovery=false` option. Default behavior is unchanged.

File: sdk/data/azcosmos/cosmos_client.go (hunk 1 above)
Comment (logangeorge01) [reply sdk/data/azcosmos/cosmos_client.go]: Good call — fixed in d89b62f. When `DisableEndpointDiscovery` is `true`, both constructors now:
- disable cross-region retries (`enableCrossRegionRetries = false`), so no failover attempt is made that would just resolve back to the client endpoint, and
- return an error if `PreferredRegions` is also set, since that combination is contradictory.

Updated the option doc + CHANGELOG to state this, and added `TestNewClientDisableEndpointDiscoveryRejectsPreferredRegions` plus a check that cross-region retries are off. Also corrected the stale "inverted" wording in the test comment (the option is propagated directly now that the internal flag is negative).

Comment (simorenoh) [reply sdk/data/azcosmos/cosmos_client.go]: This is not fully addressed yet. It is fine to accept `PreferredRegions` and document that they do not affect endpoint selection, matching Python. However, the Go client still leaves `enableCrossRegionRetries` set to `true`, so retry policies can continue cross-region failover bookkeeping even though endpoint resolution always returns the configured endpoint. Please explicitly suppress the cross-region failover paths when `DisableEndpointDiscovery` is true, while preserving intentional same-endpoint retries.

Comment (logangeorge01) [reply sdk/data/azcosmos/cosmos_client.go]: Addressed in 550214136a by separating the two concerns.

- Reverted `enableCrossRegionRetries` to stay `true` when `DisableEndpointDiscovery` is set, so same-endpoint transport retries (`maxSameRegionConnectionRetries` in `attemptRetryOnNetworkError`) keep working
- Instead, I gate only the endpoint-changing failover paths on `disableEndpointDiscovery`: the cross-region section of `attemptRetryOnNetworkError` (after the same-region budget), `attemptRetryOnEndpointFailure`, `attemptRetryOnServiceUnavailable`, `attemptRetryOnRequestTimeout` (408), and the cross-region branch of `attemptRetryOnServerError` (the in-region 5xx retry is preserved). No more `MarkEndpointUnavailable`/region-index bookkeeping on the single pinned endpoint.
- Removed the `PreferredRegions` rejection; it's now accepted and documented as having no effect on routing, matching Python. Option doc and CHANGELOG updated accordingly.

New test `TestDisableEndpointDiscoveryKeepsSameRegionRetriesButSuppressesFailover` asserts both halves: a transient failure at the pinned endpoint recovers via same-region retries, and once that budget is exhausted the call fails fast without any failover request to another advertised region. Also added `TestNewClientDisableEndpointDiscoveryAcceptsPreferredRegions` and removed the old rejection test.

File: sdk/data/azcosmos/cosmos_client.go [hunk 2]
```diff
@@ -82,12 +82,23 @@ func NewClientWithKey(endpoint string, cred KeyCredential, o *ClientOptions) (*C
...
+		disableEndpointDiscovery = o.DisableEndpointDiscovery
+	}
+	if disableEndpointDiscovery {
+		if len(preferredRegions) > 0 {
+			return nil, errors.New("azcosmos: PreferredRegions cannot be used together with DisableEndpointDiscovery; all requests are routed to the client endpoint")
+		}
+		// Discovery is disabled, so every request is pinned to the client
+		// endpoint. Cross-region retries would otherwise attempt a failover
+		// that resolves back to the same endpoint, so disable them too.
+		enableCrossRegionRetries = false
```
[the remark this answers is not in the document]
Comment (logangeorge01) [reply sdk/data/azcosmos/cosmos_client.go]: Fixed in 550214136a. I no longer set `enableCrossRegionRetries = false`; it stays `true` so the `maxSameRegionConnectionRetries` loop in `attemptRetryOnNetworkError` still runs. The two concerns are now separated — cross-region failover is gated on `disableEndpointDiscovery` in the failover paths only, so a transient DNS/connect blip at the pinned endpoint is retried in-region instead of failing immediately. Covered by `TestDisableEndpointDiscoveryKeepsSameRegionRetriesButSuppressesFailover`.

File: sdk/data/azcosmos/cosmos_client_retry_policy.go [hunk 3]
```diff
@@ -507,7 +515,7 @@ func (p *clientRetryPolicy) attemptRetryOnSessionUnavailable(isWriteOperation bo
 }
 
 func (p *clientRetryPolicy) attemptRetryOnServiceUnavailable(isWriteOperation bool, retryContext *retryContext) bool {
-	if !p.gem.locationCache.enableCrossRegionRetries || retryContext.preferredLocationIndex >= len(p.gem.preferredLocations) {
+	if !p.gem.locationCache.enableCrossRegionRetries || p.gem.locationCache.disableEndpointDiscovery || retryContext.preferredLocationIndex >= len(p.gem.preferredLocations) {
```
Comment (simorenoh) [inline sdk/data/azcosmos/cosmos_client_retry_policy.go:525]: Please apply the same `disableEndpointDiscovery` guard to `attemptRetryOnSessionUnavailable` above. A 404/1002 currently still advances session retry state and retries the pinned endpoint as though switching regions/write endpoints. Python suppresses this session-unavailable failover when endpoint discovery is disabled.

Comment (logangeorge01) [reply sdk/data/azcosmos/cosmos_client_retry_policy.go:525]: good call, pushed a fix and added a test for that case as well

**Review by simorenoh — APPROVED** [review]

File: sdk/data/azcosmos/CHANGELOG.md [hunk 4]
```diff
@@ -5,6 +5,7 @@
 ## 1.6.0-beta.4 (Unreleased)
 
 ### Features Added
+* Added `ClientOptions.DisableEndpointDiscovery`. When set to `true`, the client ignores the account's advertised readable/writable locations for routing and sends every request to the endpoint the client was constructed with. Cross-region failover is suppressed in this mode while same-region (same-endpoint) transport retries are preserved. `PreferredRegions` may still be set but has no effect on routing. This supports environments where the account advertises a document endpoint whose host is not reachable from the client's network and all traffic must instead flow through the endpoint the client was created with (e.g. a reverse proxy on a different domain).
```
Comment (tvaron3) [inline sdk/data/azcosmos/CHANGELOG.md]: add the pr link to the changelog entry

Comment (tvaron3) [inline sdk/data/azcosmos/CHANGELOG.md]: add pr link to the entry like the other entries in the changelog

Comment (logangeorge01) [reply sdk/data/azcosmos/CHANGELOG.md]: done thanks

**Review by tvaron3 — APPROVED** [review]

File: sdk/data/azcosmos/cosmos_client_retry_policy.go [hunk 5]
```diff
@@ -412,7 +420,7 @@ func (p *clientRetryPolicy) attemptRetryOnNetworkError(req *policy.Request, kind
 }
 
 func (p *clientRetryPolicy) attemptRetryOnEndpointFailure(req *policy.Request, isWriteOperation bool, retryContext *retryContext) (bool, error) {
-	if (retryContext.retryCount > maxRetryCount) || !p.gem.locationCache.enableCrossRegionRetries {
+	if (retryContext.retryCount > maxRetryCount) || !p.gem.locationCache.enableCrossRegionRetries || p.gem.locationCache.disableEndpointDiscovery {
```
[the remark this answers is not in the document]
Comment (logangeorge01) [reply sdk/data/azcosmos/cosmos_client_retry_policy.go:423]: The guarded failover paths (403 endpoint failure, 503, 408, and the cross-region 500/502/504 branch) all share the single `disableEndpointDiscovery` gate that is already exercised by the network-error and 404/1002 (session-unavailable) tests, plus `TestDisableEndpointDiscoveryKeepsSameRegionRetriesButSuppressesFailover`, which asserts the key contract: same-region retries are preserved while endpoint-changing failover is suppressed. Since these paths are guarded by the same flag rather than independent logic, the contract is covered; I'm resolving this as a non-blocking coverage nit. Happy to add the extra table-driven cases in a follow-up if the team prefers more explicit per-status coverage.

File: sdk/data/azcosmos/cosmos_global_endpoint_manager.go [hunk 6]
```diff
@@ -70,7 +70,7 @@ type updateFlight struct {
 	genAtStart uint64
 }
 
-func newGlobalEndpointManager(clientEndpoint string, pipeline azruntime.Pipeline, preferredLocations []string, refreshTimeInterval time.Duration, enableCrossRegionRetries bool) (*globalEndpointManager, error) {
+func newGlobalEndpointManager(clientEndpoint string, pipeline azruntime.Pipeline, preferredLocations []string, refreshTimeInterval time.Duration, enableCrossRegionRetries bool, enableEndpointDiscovery bool) (*globalEndpointManager, error) {
```
[the remark this answers is not in the document]
Comment (logangeorge01) [reply sdk/data/azcosmos/cosmos_global_endpoint_manager.go]: All `newGlobalEndpointManager` call sites were updated with the discovery argument in a later commit; the full package suite (`go test ./...`) compiles and is green. Resolving as addressed.

File: sdk/data/azcosmos/cosmos_location_cache.go [hunk 7]
```diff
@@ -64,14 +64,15 @@ type locationCache struct {
...
+	enableEndpointDiscovery           bool
 	locationUnavailabilityInfoMap     map[url.URL]locationUnavailabilityInfo
 	mapMutex                          sync.RWMutex
 	lastUpdateTime                    time.Time
 	enableMultipleWriteLocations      bool
 	unavailableLocationExpirationTime time.Duration
 }
 
-func newLocationCache(prefLocations []string, defaultEndpoint url.URL, enableCrossRegionRetries bool) *locationCache {
+func newLocationCache(prefLocations []string, defaultEndpoint url.URL, enableCrossRegionRetries bool, enableEndpointDiscovery bool) *locationCache {
```
[the remark this answers is not in the document]
Comment (logangeorge01) [reply sdk/data/azcosmos/cosmos_location_cache.go]: All `newLocationCache` call sites were updated to the new signature in a later commit; the package test suite compiles and passes. Resolving as addressed.

File: sdk/data/azcosmos/cosmos_location_cache.go [hunk 8]
```diff
@@ -184,6 +186,14 @@ func (lc *locationCache) resolveServiceEndpoint(locationIndex int, resourceType
 	// race with us without this lock.
 	lc.mapMutex.RLock()
 	defer lc.mapMutex.RUnlock()
+	// When endpoint discovery is disabled, ignore the account's advertised
+	// regional endpoints entirely and always route to the endpoint the client
+	// was constructed with. This keeps all traffic on the externally-routable
+	// endpoint even though the account advertises an internal document endpoint
+	// that is unreachable from this network.
+	if !lc.enableEndpointDiscovery {
+		return lc.defaultEndpoint
```
[the remark this answers is not in the document]
Comment (logangeorge01) [reply sdk/data/azcosmos/cosmos_location_cache.go]: Covered by `TestResolveServiceEndpointWithEndpointDiscoveryDisabled` (pins reads/writes to defaultEndpoint when disabled) plus the enabled-mode resolution tests. Resolving as addressed.

File: sdk/data/azcosmos/cosmos_client_options.go [hunk 9]
```diff
@@ -15,6 +15,14 @@ type ClientOptions struct {
...
 	// PreferredRegions is a list of regions to be used when initializing the client in case the default region fails.
 	PreferredRegions []string
+	// DisableEndpointDiscovery turns off the SDK's global endpoint discovery. When true,
+	// the account's advertised writable/readable locations are ignored for routing and
+	// every request is sent to the endpoint the client was constructed with. This is
+	// required in environments where the account advertises an internal document endpoint
+	// (e.g. *.docdb.azs) that is not reachable from the client's network and all traffic
+	// must instead flow through the externally-routable endpoint (e.g. via a reverse proxy).
+	// The default is false (discovery enabled), matching prior behavior.
+	DisableEndpointDiscovery bool
```
[the remark this answers is not in the document]
Comment (logangeorge01) [reply sdk/data/azcosmos/cosmos_client_options.go:31]: Constructor-level coverage added: `TestNewClientDisableEndpointDiscovery` and `TestNewClientWithKeyDisableEndpointDiscovery` build clients through the public constructors and assert routing to the client endpoint with `DisableEndpointDiscovery: true` (and default routing otherwise). Resolving as addressed.

File: sdk/data/azcosmos/cosmos_client.go [hunk 10]
```diff
@@ -82,12 +82,16 @@ func NewClientWithKey(endpoint string, cred KeyCredential, o *ClientOptions) (*C
 	}
 	preferredRegions := []string{}
 	enableCrossRegionRetries := true
+	enableEndpointDiscovery := true
 	if o != nil {
 		preferredRegions = o.PreferredRegions
+		if o.DisableEndpointDiscovery {
+			enableEndpointDiscovery = false
+		}
```
[the remark this answers is not in the document]
Comment (logangeorge01) [reply sdk/data/azcosmos/cosmos_client.go]: Addressed by `TestNewClientWithKeyDisableEndpointDiscovery` / `TestNewClientDisableEndpointDiscovery`, which exercise the public constructors (not just the cache layer) and verify data-plane requests resolve to the supplied endpoint. Resolving as addressed.

**Comment by analogrelay** [comment]
Hey @logangeorge01 , just an FYI that we're likely going to be reverting this change. In general, Cosmos cannot provide an SLA except when accessed through the regional endpoints. It's critical to our availability features.

That does **not** mean we're against providing the option for unique environments like yours, but we'd like to check in with you about your circumstances and make sure we add this functionality in a way that can clearly communicate the risks you're taking by enabling it. In your exact situation (single region, air-gapped), the risks may be well-mitigated, but we want to try and understand the scenario in a little more detail.

> This change just aligns the go sdk with dotnet which already supports disabling endpoint discovery

While this is true, it should be noted that the reason we've hesitated to add this feature to some SDKs is specifically because we've seen issues arise because of it. So, we're trying to avoid just putting it here "by default"

**Comment by analogrelay** [comment]
We'll reach out separately to see if we can figure out exactly your scenario!
